### DISCUSSION TITLE
Link module with an external library

### DISCUSSION THREAD

**Discussion by ugo94490** [nasa/fprime#976] [Q&A] [answered]
Hello,

I'm trying to link a module with an external library ``libfoo.a`` but I get the error when i run ``fprime-util build`` :
```sh
/usr/bin/ld : XComponentImpl.cpp:(.text+0x209) : undefined reference to " functionX(x, y) "
```

In order to make my module link properly I tried several methods.

First I added ``include_directories(./libs)`` so that the module knows where to find my header, then I added the option ``link_directories(./libs/)`` and ``link_libraries(foo)`` in CMake but I still get the error.

I searched the fprime directory and found the options below but I still get the error.
```cmake

set(INPUT_DEPS
./libs/libfoo.a
)

set(LINK_DEPS
-lfoo
)
```
I tried these different techniques in the CMakeList of my module and in the CMakeList at the root of the project.

I already tried what @LeStarch says in this [thread](https://groups.google.com/g/fprime-community/c/i1ED2vLly0s/m/MVjchVGnAQAJ) but it doesn't work for me.

Has anyone ever succeeded in linking a module depending on an external library?

-Ugo

**Comment by ThibFrgsGmz** [comment]
I found a project using F' that integrates a static library for one of its components : [FS](https://github.com/FP-Mission/FS/tree/dev) @lewisjaggi.

The component that uses libpng ([here](https://github.com/FP-Mission/FS/tree/dev/App/lib/lib)) is the [PiCamera](https://github.com/FP-Mission/FS/tree/dev/App/PiCamera) and here is the [cmakefile](https://github.com/FP-Mission/FS/blob/dev/App/PiCamera/CMakeLists.txt) of its component where it specifies the dependency on libpng:

```cmake
include_directories(../lib/include)
include_directories(../lib/include/libpng16)

find_library(RASPICAM raspicam /usr/local/lib)
find_library(PNG png ../lib/lib)

target_link_libraries(App_PiCamera ${RASPICAM} ${PNG})
```

I don't know if it worked in his case, I guess it did, but you can see if it helps?

Maybe you can follow up if you try to implement it.

**Reply by ugo94490** [reply]
Thank you very much for the answer!

I followed the same process as the cmake you linked but I still get the same error.

Here is what I have in the CMakeList of my module:

```cmake
set(SOURCE_FILES
  "${CMAKE_CURRENT_LIST_DIR}/xxxComponentAi.xml"
  "${CMAKE_CURRENT_LIST_DIR}/yyySerializableAi.xml"
  "${CMAKE_CURRENT_LIST_DIR}/xxxComponentImpl.cpp"
)

register_fprime_module()

include_directories(/path/to/foo.h)
find_library(FOO foo /path/to/foo.a)
target_link_libraries(FOLDER_xxxx ${FOO})
```

**Reply by astroesteban** [reply]
The modern CMake way of doing this would be:
```cmake
    # The imported library's scope is limited to the directory it's imported in. The optional GLOBAL keyword extends that to the global scope 
    add_library(foo STATIC IMPORTED GLOBAL)
    set_target_properties(foo PROPERTIES IMPORTED_LOCATION /path/to/foo.a)
    # Following is optional if your library has header dependencies
    target_include_directories(foo INTERFACE /path/to/headers)
    target_link_libraries(FOLDER_xxxx PRIVATE foo)
```
This way you make `foo` a proper CMake target.

**Reply by lewisjaggi** [reply]
Hello, i confim, it works for us 👍🏻.

Hint : you must not put the filename at the end in : 

```cmake
include_directories(/path/to/foo.h)
find_library(FOO foo /path/to/foo.a)
```

You can try :

```cmake
include_directories(/path/to)
find_library(FOO foo /path/to)
```

The second parameter "foo" refers to libfoo.a

**Reply by LeStarch** [reply]
I just wanted to point out that you should make sure to have the fprime module name.  You can do that in two ways:

1. Convert the path from the root of your project to underscores.  i.e.  Fw/Types becomes Fw_Types
2. Use the `get_module_name()` function.  It performs step 1 automatically and is used to generate the target name to begin with
```
get_module_name("${CMAKE_CURRENT_LIST_DIR}")
target_link_libraries("${MODULE_NAME}" ${FOO})
```

If it is a unit test target (e.g. added with `fprime_register_ut` then add "_ut" to the end.

**Reply by ugo94490** [reply]
Hello,

Thank you very much for your answers. I have tried the different methods you have advised.

Here is what I have now in the CMakeLists.txt in my module :
```cmake
register_fprime_module()

get_module_name("${CMAKE_CURRENT_LIST_DIR}")

include_directories(path/to/)
find_library(FOO _foo /path/to) # _foo because my library is called lib_foo.a

target_link_libraries(${MODULE_NAME} ${FOO})
```

I still have the same error when linking to the binary : 
```
/usr/bin/ld : XComponentImpl.cpp:(.text+0x209) : undefined reference to " functionX(x, y) "
```

Nevertheless, ``get_module_name`` seems to work.

**Reply by ThibFrgsGmz** [reply]
I will tackle your problem and gather the above information. Depending on my availability and if I can, I will put the full procedure here so that we are in sync on the full process to follow.

**Reply by ThibFrgsGmz** [reply]
OK I may have found the source of your problem after reproducing it: you may be using a C library that doesn't have the famous ``external "c"`` tag in its header file.

If so, I solved it with @lewisjaggi and @LeStarch's method, but I still have the same problem if I use @astroesteban's method.

And here are the explanations that I could gather:

C and C++ languages use different techniques to handle function names.
Unlike C, C++ has built-in function overloading, which means that the linker must modify the function name to indicate the specific prototype it should call.

On the other hand, declaring a function with ``external "c"`` changes the linking requirements so that the C++ compiler does not add the extra processing information to the symbol.
[ref](https://embeddedartistry.com/blog/2017/05/01/mixing-c-and-c-extern-c/)

---

I will put in another post a short summary of the procedure to gather the information, even if it doesn't solve your problem, it might help others if it is easily visible.

**Reply by astroesteban** [reply]
With my approach what happens if you change the `target_link_libraries(FOLDER_xxxx PRIVATE foo)` part to `target_link_libraries(FOLDER_xxxx PUBLIC foo)`?

**Reply by ThibFrgsGmz** [reply]
@astroesteban  same result, in my case both versions break the generate chain of the command ``fprime-util generate``.

According to the ``CMakeError.log`` the linker has lost the references of the pthread library:

```cmake
Run Build Command(s):/usr/bin/make cmTC_8ec74/fast && /usr/bin/make -f CMakeFiles/cmTC_8ec74.dir/build.make CMakeFiles/cmTC_8ec74.dir/build
make[1]: Entering directory '/home/ThibFrgsGmz/Documents/CODES/fprime-poc/build-fprime-automatic-native/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_8ec74.dir/src.c.o
/usr/bin/cc   -DCMAKE_HAVE_LIBC_PTHREAD   -o CMakeFiles/cmTC_8ec74.dir/src.c.o   -c /home/ThibFrgsGmz/Documents/CODES/fprime-poc/build-fprime-automatic-native/CMakeFiles/CMakeTmp/src.c
Linking C executable cmTC_8ec74
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_8ec74.dir/link.txt --verbose=1
/usr/bin/cc  -DCMAKE_HAVE_LIBC_PTHREAD    CMakeFiles/cmTC_8ec74.dir/src.c.o  -o cmTC_8ec74 
/usr/bin/ld: CMakeFiles/cmTC_8ec74.dir/src.c.o: in function `main':
src.c:(.text+0x46): undefined reference to `pthread_create'
/usr/bin/ld: src.c:(.text+0x52): undefined reference to `pthread_detach'
/usr/bin/ld: src.c:(.text+0x63): undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
make[1]: *** [CMakeFiles/cmTC_8ec74.dir/build.make:87: cmTC_8ec74] Error 1
make[1]: Leaving directory '/home/ThibFrgsGmz/Documents/CODES/fprime-poc/build-fprime-automatic-native/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_8ec74/fast] Error 2
```
Here is what I had put in my cmake file to link my libQuaternion.a library located in fprime-poc/libs (with its header file):

```cmake
add_library(Quaternion STATIC IMPORTED GLOBAL)
set_target_properties(Quaternion PROPERTIES IMPORTED_LOCATION "${PROJECT_SOURCE_DIR}/libs")
target_include_directories(Quaternion INTERFACE "${PROJECT_SOURCE_DIR}/libs")
target_link_libraries("${MODULE_NAME}" PUBLIC Quaternion)
# Or 
# target_link_libraries("${MODULE_NAME}" PRIVATE Quaternion)
```

If I comment out these lines before I run the ``fprime-util generate`` command, the generation works. And then if I put them back just before running ``fprime-util build``, I get the ``undefined reference`` error.

Maybe I made a mistake somewhere?

**Reply by astroesteban** [reply]
Interesting. I don't know why it would lose pthreads. Maybe pthread needs to get linked against the `Quaternion` target but your fix worked so I'd go with your approach.

**Comment by ThibFrgsGmz** [comment, marked as answer]
If you tried to link an external C library, make sure you have the ``external "c"`` tag so that the C++ compiler does not add extra manipulation information to the function name symbols.

Then in the ``CMakeLists.txt`` of the component in which you want to use the external library you must add the following lines. In my case I try to link a quaternion calculation library (named libQuaternion.a).

```cmake
# Retrieve the name of the module that converts the root path of your project to a underscore, i.e. Fw/Types becomes Fw_Types.
get_module_name("${CMAKE_CURRENT_LIST_DIR}")

# Tell the compiler the path to the library header file.
include_directories("${PROJECT_SOURCE_DIR}/libs")

# Find all the libraries contained in the directory you specified.
find_library(LIBQUATERNION Quaternion "${PROJECT_SOURCE_DIR}/libs")
# Note: in my case the LIBQUATERNION equals /home/ThibFrgsGmz/Documents/CODES/fprime-poc/libs/libQuaternion.a

# Specify to the linker the libraries to use when linking the given component and its dependents. 
target_link_libraries("${MODULE_NAME}" ${LIBQUATERNION})
```

If I remove the ``external "c"`` tag from the library header file, I get the same error as you, if I put it back, the error disappears.

Hopefully this solves your problem.

**Reply by ugo94490** [reply]
Hello,

Thank you very much for the help, my problem was due to the ``external "C"`` that I had forgotten to add to my lib.

Thank you all for taking the time to help me.
